tests/integration/utils.py

```diff
@@ -347,21 +347,22 @@ def assert_expected_files(source_path, expected_files, tmpdir):
             expected_archive = os.path.join(tmpdir, f"archive_{dir_identifier}.tar.gz")
             # A directory path with extracted expected data
             expected_data_path = os.path.join(tmpdir, f"expected_data_{dir_identifier}")
-            # Dependencies instead of source code are saved as archives.
-            # Therefore we should extract it firstly
-            if os.path.isdir(test_path):
-                package_root_dir = test_path
-            else:
-                archive_path = test_path
-                shutil.unpack_archive(archive_path, deps_data_path)
-                # deps_data_path is unique and contains only one expected package
-                package_root_dir = os.path.join(deps_data_path, os.listdir(deps_data_path)[0])
+
+            package_root_dir = unpack_test_archives(deps_data_path, test_path)
+
             download_archive(archive_url, expected_archive)
+            if os.path.exists(expected_data_path):
+                shutil.rmtree(expected_data_path)
```

Comment:
Weirdly enough, without these two lines RubyGems cached dependency test would fail. But with these lines, I'm confident cached dependencies tests pass (tested with Pip and RubyGems), so we can either merge it as it is or I can look into it more.

Reply:
The weird part is that these files should be deleted later in this function and they maybe are, but then when this function is called for the second request, it raises permission error while trying to unpack the archive, as if they still existed. And deleting them this way solves the issue.

Reply:
Oh, I think I found the bug /facepalm
```python
            for temp_data in [deps_data_path, expected_data_path, expected_archive]:
                if os.path.exists(deps_data_path):
                                  --------------
                    shutil.rmtree(temp_data)
```
So the tests never delete `expected_data_path` or `expected_archive` :sweat_smile:

Reply:
Well, I have no idea what I was doing when I was debugging it :sweat_smile: